pydantic/functional_validators.py

```diff
@@ -409,32 +409,17 @@ def __call__(  # noqa: D102
         ...
 
 
-class ModelAfterValidatorWithoutInfo(Protocol):
-    """A `@model_validator` decorated function signature. This is used when `mode='after'` and the function does not
-    have info argument.
-    """
-
-    @staticmethod
-    def __call__(  # noqa: D102
-        self: _ModelType,  # type: ignore
-    ) -> _ModelType:
-        ...
-
-
-class ModelAfterValidator(Protocol):
-    """A `@model_validator` decorated function signature. This is used when `mode='after'`."""
-
-    @staticmethod
-    def __call__(  # noqa: D102
-        self: _ModelType,  # type: ignore
-        __info: _core_schema.ValidationInfo,
-    ) -> _ModelType:
-        ...
+ModelAfterValidatorWithoutInfo = Callable[[_ModelType], _ModelType]
+"""A `@model_validator` decorated function signature. This is used when `mode='after'` and the function does not
+have info argument.
+"""
```

Comment:
If I understand correctly, mypy didn't recognize the the previously annotated `Protocol` object as `callable`? (Even though it had `call` method? By switching to explicitly annotating with `Callable`, the annotations are correct?

Reply:
I like that we're annotating callable functions as callable though. (Though `Protocol` is supposed to solve this for us, no?)

Reply:
In theory you can use either, but it's a little bit weird because here we want to require the `self` argument, in particular because it helps distinguish from the other alternative signatures. But my hack to make that work does not seem to have been compatible with mypy so this is a simpler albeit less pedantic version that hopefully will be.

Reply:
> here we want to require the `self` argument, in particular because it helps distinguish from the other alternative signatures

I don't know if I'm understanding your thought correctly.
But if those validators work with other signatures as well (e.g. with classmethod signatures) it would be misleading imo. The implementation as far as I saw supports static-, class- and instance-signatures and even decorators. Therefore, I don't think you have to enforce an instance-signature here (i.e. with self argument).

Actually, even the [documentation](https://docs.pydantic.dev/latest/usage/validators/#model-validators) suggests a classmethod-signature for `model_validator` with `mode="after"`. Does the solution of this PR work with classmethod-signatures though, did you test it?

Reply:
> Actually, even the [documentation](https://docs.pydantic.dev/latest/usage/validators/#model-validators) suggests a classmethod-signature for model_validator with mode="after"

This is precisely what we wanted to avoid, the docs are wrong.

Reply:
Ok, I tried around by myself but I really couldn't get a Protocol typing for instance methods to work together with the decorator. At least with mypy. Very sad, maybe it's possible in the future.

Btw, why do you even want to avoid classmethod signatures? Is there anywhere a discussion about this?

Reply:
> Btw, why do you even want to avoid classmethod signatures? Is there anywhere a discussion about this?

You're getting an instance of the model. Of course `(cls: Type['Model'], model: 'Model')` works but why do that instead of `(self)` which works great with type checkers, etc.